flake.nix

```diff
@@ -0,0 +1,131 @@
+{
+  description =
+    "Javascript implementation of the Bitcoin protocol for any Bitcoin based coins";
+
+  # Nixpkgs / NixOS version to use.
+  inputs.nixpkgs.url = "nixpkgs/nixos-unstable"; # needed for nodePackage.terser
+
+  # Upstream source tree(s).
+  inputs.bitcoin-transactions-src = {
+    url = "path:./.";
+    flake = false;
+  };
+
+  outputs = { self, nixpkgs, bitcoin-transactions-src }:
+
+    let
+      # System types to support.
+      supportedSystems = [ "x86_64-linux" ];
+
+      # Helper function to generate an attrset '{ x86_64-linux = f "x86_64-linux"; ... }'.
+      forAllSystems = f:
+        nixpkgs.lib.genAttrs supportedSystems (system: f system);
+
+      # Nixpkgs instantiated for supported system types.
+      nixpkgsFor = forAllSystems (system:
+        import nixpkgs {
+          inherit system;
+          overlays = [ self.overlay ];
+        });
+
+    in {
+
+      overlay = final: prev: {
+        bitcoin-transactions =
+          let inherit (final) stdenv nodejs lib nodePackages;
+          in stdenv.mkDerivation {
+            name = "bitcoin-transactions";
+
+            src = bitcoin-transactions-src;
+
+            buildInputs = [ nodejs ];
+
+            dontBuild = true;
+
+            installPhase = ''
+              mkdir -p $out/
+              cp -r ./* $out/
+            '';
+
+            meta = {
+              homepage = "https://peersm.com/wallet";
+              description =
+                "Javascript implementation of the Bitcoin protocol for any Bitcoin based coins";
+              license = lib.licenses.mit;
+            };
+          };
+      };
+
+      # Provide some binary packages for selected system types.
+      packages = forAllSystems
+        (system: { inherit (nixpkgsFor.${system}) bitcoin-transactions; });
+
+      # These apps load the package as a standalone offline webapp
+      # For these to work the XDG browser variable needs to be set to something that exists
+      # To test run `xdg-settings get default-web-browser` if this fails
+      # prepend the nix run command with your installed browser with
+      # BROWSER=<installed browser> nix run ...
+      apps = forAllSystems (system: {
+        wallet = let pkgs = nixpkgsFor.${system};
+        in {
+          type = "app";
+          program = "${
+              pkgs.writeShellScriptBin "bitcoin-transactions" ''
+                ${pkgs.xdg-utils}/bin/xdg-open ${pkgs.bitcoin-transactions}/html/wallet.html
+              ''
+            }/bin/bitcoin-transactions";
+        };
+        # This shows the Javascript unminified and readable
+        wallet-clear = let pkgs = nixpkgsFor.${system};
```

Comment:
This might be overkill, but since the program here is the same for both, you could make a function `genProgram` which takes in a parameter specifying `walletclear` or `wallet` and generates the program based on that.

Reply:
What is the advantage to writing a nix function to generate a program as opposed to adding flags to the `ShellScriptBin` like
```
          program = "${
              pkgs.writeShellScriptBin "bitcoin-transactions" ''
                # Clear shows the Javascript unminified and readable
                if [ "$1" == "clear" ] || [ "$1" == "c" ]; then
                  ${pkgs.xdg-utils}/bin/xdg-open ${pkgs.bitcoin-transactions}/html/walletclear.html
                else
                  ${pkgs.xdg-utils}/bin/xdg-open ${pkgs.bitcoin-transactions}/html/wallet.html
                fi
              ''
            }/bin/bitcoin-transactions";
```

Which could then be called with `nix run .#wallet -- c` etc.

Reply:
Ooh, yeah. I like this more. +1. I think only the path would need to be set (since the invocation is the same for both).
```bash
          program = "${
              pkgs.writeShellScriptBin "bitcoin-transactions" ''
                # Clear shows the Javascript unminified and readable
                if [ "$1" == "clear" ] || [ "$1" == "c" ]; then
                  PAGE=html/walletclear.html
                else
                  PAGE=html/wallet.html
                fi
                ${pkgs.xdg-utils}/bin/xdg-open ${pkgs.bitcoin-transactions}/$PAGE
              ''
            }/bin/bitcoin-transactions";
```

Reply:
Nice catch ill make that change now :+1: